modules/assim.sequential/R/downscale_function.R

```diff
@@ -114,62 +153,70 @@ SDA_downscale_preprocess <- function(data_path, coords_path, date, carbon_pool)
-  predictors <- as.data.frame(terra::extract(covariates, site_coordinates, ID = FALSE))
-  
+
+  # Convert site coordinates to an sf object using the helper function
+  site_coordinates_sf <- .convert_coords_to_sf(preprocessed$site_coordinates)
+
+  # Extract predictors from covariates <U+2013> support both SpatRaster and sf objects:
+  if (inherits(covariates, "SpatRaster")) {
+    # For raster objects, convert the sf coordinates to SpatVector before extraction
+    predictors <- as.data.frame(terra::extract(covariates, terra::vect(site_coordinates_sf), ID = FALSE))
+  } else if (inherits(covariates, "sf")) {
+    # For sf objects, spatial join
+    predictors <- sf::st_drop_geometry(sf::st_join(site_coordinates_sf, covariates, join = sf::st_intersects))
+  } else {
+    stop("Unsupported covariates object. Must be a SpatRaster or an sf object.")
+  }
   # Dynamically get covariate names
   covariate_names <- names(predictors)
-  
+
   # Create a single data frame with all predictors and ensemble data
   full_data <- cbind(carbon_data, predictors)
-  
+
   # Split the observations into training and testing sets
   if (!is.null(seed)) {
-    set.seed(seed)  # Only set seed if provided
+    set.seed(seed) # Only set seed if provided
   }
-  sample <- sample(1:nrow(full_data), size = round(0.75 * nrow(full_data)))
+  sample <- sample(seq_len(nrow(full_data)), size = round(0.75 * nrow(full_data)))
   train_data <- full_data[sample, ]
   test_data <- full_data[-sample, ]
-  
+
   # Prepare data for both RF and CNN
   x_data <- as.matrix(full_data[, covariate_names])
   y_data <- as.matrix(carbon_data)
-  
+
   # Calculate scaling parameters from all data
   scaling_params <- list(
     mean = colMeans(x_data),
     sd = apply(x_data, 2, stats::sd)
   )
-  
+
   # Normalize the data
   x_data_scaled <- scale(x_data, center = scaling_params$mean, scale = scaling_params$sd)
-  
+
   # Split into training and testing sets
   x_train <- x_data_scaled[sample, ]
   x_test <- x_data_scaled[-sample, ]
   y_train <- y_data[sample, ]
   y_test <- y_data[-sample, ]
-  
+
   # Initialize lists for outputs
   models <- list()
   maps <- list()
   predictions <- list()
-  
+
   if (model_type == "rf") {
     for (i in seq_along(carbon_data)) {
       ensemble_col <- paste0("ensemble", i)
       formula <- stats::as.formula(paste(ensemble_col, "~", paste(covariate_names, collapse = " + ")))
       models[[i]] <- randomForest::randomForest(formula,
-                                                data = train_data,
-                                                ntree = 1000,
-                                                na.action = stats::na.omit,
-                                                keep.forest = TRUE,
-                                                importance = TRUE)
-      
+        data = train_data,
+        ntree = 1000,
+        na.action = stats::na.omit,
+        keep.forest = TRUE,
+        importance = TRUE
+      )
+
```

Comment:
Another suggestion I've made in the past is that there may be logical/computational advantages to separating the function that builds the downscaling models and a separate function that makes the predictions to a map. Right now one can't just fit the model and evaluate the results without also running the (more computationally costly) mapping